document-legacy/src/document.rs

```diff
@@ -559,50 +621,156 @@ impl Document {
+						DocumentChanged,
+						CreatedLayer {
+							path: updated_layer_path,
+							is_selected: false,
+						},
+						FolderChanged { path: destination_path.clone() },
+					]);
+				} else if !duplicating {
+					let folder = self.folder_mut(folder_path)?;
+					folder.add_layer(*layer, Some(layer_id), insert_index).ok_or(DocumentError::IndexOutOfBounds)?;
+
+					responses.extend([
+						DocumentChanged,
+						CreatedLayer {
+							path: destination_path.clone(),
+							is_selected: true,
+						},
+						FolderChanged { path: folder_path.to_vec() },
+					]);
+				}
+
+				responses.extend(update_thumbnails_upstream(&destination_path));
 				self.mark_as_dirty(&destination_path)?;
 
-				fn aggregate_insertions(folder: &FolderLayer, path: &mut Vec<LayerId>, responses: &mut Vec<DocumentResponse>) {
+				// Recursively iterate through each layer in a folder and add it to the responses vector
+				fn aggregate_insertions(folder: &FolderLayer, path: &mut Vec<LayerId>, responses: &mut Vec<DocumentResponse>, duplicating: bool) {
 					for (id, layer) in folder.layer_ids.iter().zip(folder.layers()) {
 						path.push(*id);
-						responses.push(DocumentResponse::CreatedLayer { path: path.clone() });
+
+						responses.push(DocumentResponse::CreatedLayer {
+							path: path.clone(),
+							is_selected: !duplicating,
+						});
 						if let LayerDataType::Folder(f) = &layer.data {
-							aggregate_insertions(f, path, responses);
+							aggregate_insertions(f, path, responses, duplicating);
 						}
+
 						path.pop();
 					}
 				}
 
-				let mut responses = Vec::new();
 				if let Ok(folder) = self.folder(&destination_path) {
-					aggregate_insertions(folder, &mut destination_path.as_slice().to_vec(), &mut responses)
+					aggregate_insertions(folder, &mut destination_path.as_slice().to_vec(), &mut responses, duplicating);
 				};
 
-				responses.extend([DocumentChanged, CreatedLayer { path: destination_path.clone() }, FolderChanged { path: folder_path.to_vec() }]);
-				responses.extend(update_thumbnails_upstream(&destination_path));
 				Some(responses)
 			}
 			Operation::DuplicateLayer { path } => {
+				// Notes for review: I wasn't sure on how to apply unwrap_or() to lines of code that use the funciton self.layer()
+
 				let layer = self.layer(&path)?.clone();
 				let (folder_path, _) = split_path(path.as_slice()).unwrap_or((&[], 0));
-				let folder = self.folder_mut(folder_path)?;
 
+				let layer_is_folder = layer.as_folder().is_ok();
+
+				// Recursively collect each of the nested folders and shapes if the layer is a folder
+				let mut duplicated_layers = if layer_is_folder { recursive_collect(self, &path) } else { Vec::new() };
+				// Duplicate the Layer objects for our duplicated layers
+				// Sort both vectors by the layer path size, from shallowest to deepest
+				let duplicated_layers_objects: Vec<Layer> = duplicated_layers.iter().map(|p| self.layer(p.as_slice()).unwrap()).cloned().collect();
```

Comment:
i see you had concerns about how to handle the unwrap on self.layer, maybe using the ? notation to handle the Result type that self.layer returns. Or at least an unwrap_or so we dont crash

Reply:
@Keavon Me and ollie met on discord to find a way to rewrite the code without the unwrap() function, but couldn't figure it out. When you review the code again, can you see if you have any solutions for this.

Reply:
You can do something like:
```rs
let duplicated_layers_objects = duplicated_layers.iter().map(|p| self.layer(p).cloned()).collect::<Result<Vec<Layer>, DocumentError>>()?;
```
This works because collecting an iterator of `Result<T, E>` allows you to create a `Result<Vec<T>, E>`. For more information about collect, see the [collect function docs](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.collect).